src/components/ui4/ButtonUi4.tsx

```diff
@@ -0,0 +1,154 @@
+
+interface Props {
+  children?: React.ReactNode
+
+  // Called when the user presses the button.
+  // If the callback returns a promise, the button will disable itself
+  // and show a spinner until the promise resolves.
+  onPress?: () => void | Promise<void>
+
+  // Whether to center the button or stretch to fill the screen.
+  // Defaults to 'auto', letting the parent component be in charge:
+  alignSelf?: 'auto' | 'stretch' | 'center'
+
+  // True to dim the button & prevent interactions:
+  disabled?: boolean
+
+  // If this is set, the component will insert a text node after its children:
+  label?: string
+
+  // The gap around the button. Takes 0-4 numbers (top, right, bottom, left),
+  // using the same logic as the web `margin` property. Defaults to 0.
+  marginRem?: number[] | number
+
+  // True to show a spinner after the contents:
+  spinner?: boolean
+
+  // Which visual style to use. Defaults to primary (solid):
+  type?: ButtonTypeUi4
+}
+
+/**
+ * A stand-alone button to perform the primary action in a modal or scene.
+ */
+export function ButtonUi4(props: Props) {
+  const { alignSelf = 'center', children, disabled = false, label, marginRem, onPress, type = 'primary', spinner = false } = props
+
+  // `onPress` promise logic:
+  const [pending, handlePress] = usePendingPress(onPress)
+
+  // Styles:
+  const theme = useTheme()
+  const styles = getStyles(theme)
+
+  let buttonShadow, spinnerColor, textStyle, gradientProps
+  if (type === 'primary') {
+    textStyle = styles.primaryText
+    buttonShadow = styles.primaryShadow
+
+    spinnerColor = theme.buttonPrimaryUi4.spinnerColor
+    gradientProps = theme.buttonPrimaryUi4.gradientProps
+  } else if (type === 'secondary') {
+    textStyle = styles.secondaryText
+    buttonShadow = styles.secondaryShadow
+
+    spinnerColor = theme.buttonSecondaryUi4.spinnerColor
+    gradientProps = theme.buttonSecondaryUi4.gradientProps
+  } else {
+    // (type === 'tertiary')
+    textStyle = styles.tertiaryText
+    buttonShadow = styles.tertiaryShadow
+
+    spinnerColor = theme.buttonTertiaryUi4.spinnerColor
+    gradientProps = theme.buttonTertiaryUi4.gradientProps
+  }
+
+  const dynamicGradientStyles = {
+    alignSelf,
+    opacity: disabled ? 0.3 : pending ? 0.7 : 1,
+    ...sidesToMargin(mapSides(fixSides(marginRem, 0), theme.rem))
+  }
+  return (
+    <TouchableOpacity disabled={disabled || pending} style={[buttonShadow, styles.borderRadiusCommon]} onPress={handlePress}>
+      <LinearGradient {...gradientProps} style={[styles.gradientLayoutCommon, dynamicGradientStyles, styles.borderRadiusCommon]}>
+        {pending ? null : children}
+        {pending || label == null ? null : (
+          <Text adjustsFontSizeToFit minimumFontScale={0.75} numberOfLines={1} style={textStyle}>
+            {label}
+          </Text>
+        )}
+        {!pending && !spinner ? null : <ActivityIndicator color={spinnerColor} style={styles.spinnerCommon} />}
```

Comment:
Can we use this opportunity to fix an old `MainButton` mistake? Can we make `spinner` work the same as `pending`? So it's always `(pending || spinner) ...` for each of these three conditions. That way, passing `spinner={true}` as a prop would achieve the same result as having `onPress` return a promise.